feat: Add support for Kafka clusters sharding

olksdr opened the pull request:
This allows to configured many different Kafka clusters per topic. The
basic idea is to provide the numebr of shards and the lower bound for
the range with the kafka config name and the topic name on that cluster

```
 metrics:
     shards: 65000
     mapping:
         0:
             name: "ingest-metrics-1"
             config: "metric_1"
         25000:
             name: "ingest-metrics-2"
             config: "metrics_2"
         45000:
             name: "ingest-metrics-3"
             config: "metrics_3"
```

This will also allow to configure one Kafka clusters and use different
topics names on it.

olksdr committed:
feat: Add support for Kafka clusters sharding

This allows to configured many different Kafka clusters per topic. The
basic idea is to provide the numebr of shards and the lower bound for
the range with the kafka config name and the topic name on that cluster

```
 metrics:
     shards: 65000
     mapping:
         0:
             name: "ingest-metrics-1"
             config: "metric_1"
         25000:
             name: "ingest-metrics-2"
             config: "metrics_2"
         45000:
             name: "ingest-metrics-3"
             config: "metrics_3"
```

This will also allow to configure one Kafka clusters and use different
topics names on it.

INGEST-1592

olksdr committed:
ref: Propagate number of shards, remove some duplucation

olksdr committed:
ref: Enable hashing and thread topic name though the kafka config

olksdr committed:
Merge branch 'master' into feat/kafka-sharding-config

olksdr committed:
Merge branch 'master' into feat/kafka-sharding-config

olksdr committed:
ref: Add sharded producers to outcome actor

olksdr committed:
feat: Add changelog entry

olksdr committed:
feat: use either org_id or project_id for sharding in outcomes when configured

olksdr committed:
test: Add integration tests for checking the kafka sharding

olksdr committed:
feat: fail to start if the sharding configuration is invalid

olksdr committed:
Merge branch 'master' into feat/kafka-sharding-config

flub commented on relay-config/src/config.rs:
```diff
@@ -882,41 +885,106 @@ pub enum TopicAssignment {
+pub struct Sharded {
+    /// The number of shards used for this topic.
+    shards: u64,
+    /// The Kafka configuration assigned to the specific shard range.
```
could you describe that the u64 is the start index and the next u64 describes the range.  Explicitly calling out what's inclusive (i'm assuming the start u64 is inclusive, the end u64 is excluded and part of the next range).  Maybe even write an example out on the struct doc comment like you did in the PR description.

olksdr replied:
@flub, please, have a look into https://github.com/getsentry/relay/pull/1454/commits/f9679331735dcfa7c8a490bcc6922e69f2a0d0f9 if this is something you had in mind?

flub replied:
yep, 💯

flub commented on relay-server/src/actors/outcome.rs:
```diff
@@ -835,17 +820,46 @@ impl OutcomeProducerService {
+                let org_id_hash = hasher.finish();
+                let shard = org_id_hash % shards;
+
+                // should be ok to unwrap since we MUST have at least one range defined
```
This is reasonable but because it is so far apart from each other in the code I'm a bit uncomfortable about it.  One way would be make a custom type for the producers, then it is enforced in that type and relying on that invariant would be a lot nicer and keeps the code interacting with the invariant close together.  I guess this would be a newtype over the BTreeMap with a few small methods.

olksdr replied:
That's a very good point. I will look into it. 
It might also simplify the code and removes some duplication.

olksdr replied:
So, I moved the same functionality under the `Sharded` producer in store actore [d4a05c2](https://github.com/getsentry/relay/pull/1454/commits/d4a05c23f39c9238481d9338b0b6b7f73ebb7010). 
I do not know if it's a good idea to create a new type over `BTreeMap`, which can be a bit more complicated to understand.
But I might have to pick your brain on this , and how to implement better.

flub commented on relay-server/src/actors/store.rs:
```diff
@@ -58,33 +58,127 @@ fn make_distinct_id(s: &str) -> Uuid {
+                let mut hasher = FnvHasher::default();
+                std::hash::Hash::hash(&organization_id, &mut hasher);
+                let org_id_hash = hasher.finish();
+                let shard = org_id_hash % shards;
```
this code exists in two places already, given the nature of it - i suspect both would need updating together all the time - it might be better to create a method somewhere.  Maybe also on the newtype around the BTreeMap I already suggested.

olksdr replied:
moved this into the impl for `Sharded` producer in  [d4a05c2](https://github.com/getsentry/relay/pull/1454/commits/d4a05c23f39c9238481d9338b0b6b7f73ebb7010)

olksdr committed:
Merge branch 'master' into feat/kafka-sharding-config

olksdr committed:
feat: Add better docs for the new config foramt

olksdr committed:
ref: Remove some code duplication, and use same functionality under Sharded store producer

jjbayer commented on tests/integration/fixtures/relay.py:
```diff
@@ -133,7 +134,7 @@ def inner(
 
         dir = config_dir("relay")
-        dir.join("config.yml").write(json.dumps(default_opts))
+        dir.join("config.yml").write(yaml.dump(default_opts))
```
Why is this change necessary?

olksdr replied:
There are few things here:
* first of all I think we should use Yaml for testing the configuration, since we use Yaml config file to configure the service (not JSON) - so we can test this as closer to real world 
* secondly, and which is more relevant to this change - that with new additions to the configuration we have to have the integers as the keys in the map, but JSON can only have strings there. 

And here is the thing, we if want to also support JSON we will have to relay on parsing the strings, maybe in addition to numbers (?),  and getting integers from there.  What do you think?

jjbayer replied:
The integer keys make sense. To maintain JSON compatibility, we could also change the format to a list (like below), but then we would have to validate that there are no duplicate entries. So I think sticking with yaml makes sense.

```yaml
ranges:
    - start_at: 0
      name: "ingest-metrics-1"
      config: "metrics_1"
    - start_at: 25000
      name: "ingest-metrics-2"
      config: "metrics_2"
```

jjbayer commented on relay-server/src/actors/store.rs:
```diff
@@ -58,33 +58,143 @@ fn make_distinct_id(s: &str) -> Uuid {
+                    .send(record)
+                    .map_err(|(kafka_error, _message)| StoreError::SendFailed(kafka_error))?;
+            }
+        }
```
There's code duplication in here, might be nice to only use the `match` statement to obtain the topic name & producer:

```suggestion
        let (topic_name, producer) = match self {
            Producer::Single {
                topic_name,
                producer,
            } => (topic_name.as_str(), producer),
            Producer::Sharded(sharded) => sharded.get_producer(organization_id),
        };

        let record = BaseRecord::to(topic_name).key(&key).payload(&serialized);
        producer
            .send(record)
            .map_err(|(kafka_error, _message)| StoreError::SendFailed(kafka_error))?;

```

jjbayer commented on relay-server/src/actors/store.rs:
```diff
@@ -58,33 +58,143 @@ fn make_distinct_id(s: &str) -> Uuid {
-    for config_p in kafka_config {
-        client_config.set(config_p.name.as_str(), config_p.value.as_str());
+/// Sharded producer configuration.
+pub struct Sharded {
```
nit: Should we call this `ShardedProducer`?

olksdr replied:
Yeah, this is must better name!  I will make a change.

jjbayer commented on relay-server/src/actors/store.rs:
```diff
@@ -58,33 +58,143 @@ fn make_distinct_id(s: &str) -> Uuid {
+
+impl Sharded {
+    /// Returns topic name and the Kafka producer based on the provided sharding key.
+    pub fn get_producer(&self, sharding_key: u64) -> (&str, Arc<ThreadedProducer>) {
```
Might be a premature generalization, but we could make this function generic:
```suggestion
    pub fn get_producer(&self, sharding_key: impl Hash) -> (&str, Arc<ThreadedProducer>) {
```

jjbayer commented on relay-server/src/actors/store.rs:
```diff
@@ -58,33 +58,143 @@ fn make_distinct_id(s: &str) -> Uuid {
+        let mut hasher = FnvHasher::default();
+        std::hash::Hash::hash(&sharding_key, &mut hasher);
+        let org_id_hash = hasher.finish();
+        let shard = org_id_hash % self.shards;
```
Since `org_id` is not in the function signature, I would also rename this variable:
```suggestion
        let hashed = hasher.finish();
        let shard = hashed % self.shards;
```

jjbayer commented on relay-config/src/config.rs:
```diff
@@ -882,41 +885,126 @@ pub enum TopicAssignment {
+        /// The maximum number of logical shards for this set of configs.
+        shards: u64,
+        /// The list of the sharded Kafka configs.
+        configs: BTreeMap<u64, KafkaParams<'a>>,
```
We have this mapping in three different places now, I wonder if it would be possible to have it only once.

olksdr replied:
@jjbayer I've tried to remove one of the levels of indirection to eliminate of of the repeating BTreeMap, see [0db194e](https://github.com/getsentry/relay/pull/1454/commits/0db194e5b2a53a2cd8462d8dc369c997a38f6439) 
Please, have another look if this is somewhat better approach.

jjbayer replied:
Yes, I think it makes sense this way.

olksdr committed:
ref: Address the review comments: remove code duplication

olksdr committed:
ref: Remove one level of inderections

Now Kafka config with topic name and parameters is produced from
TopicAssignment enum

jjbayer commented on relay-config/src/config.rs:
```diff
@@ -2236,11 +2200,5 @@ processing:
-        assert!(matches!(
-            kafka_config_profiles,
-            KafkaConfigName::Single { .. }
-        ));
```
Any way we could keep this assertion?

olksdr replied:
done in [1c3d5e0](https://github.com/getsentry/relay/pull/1454/commits/1c3d5e07e5a6d8352b6fadbed57a04d649374f0f).

jjbayer commented on relay-config/src/config.rs:
```diff
@@ -985,27 +968,56 @@ impl From<String> for TopicAssignment {
-                topic_name: topic.as_str(),
-                kafka_config_name: None,
-            },
+    fn kafka_config_name<'a>(
```
Should we update the doc comment and the function name, now that this function returns a full config rather than a name?

olksdr replied:
Good point, done in [1c3d5e0](https://github.com/getsentry/relay/pull/1454/commits/1c3d5e07e5a6d8352b6fadbed57a04d649374f0f)

olksdr committed:
Merge branch 'master' into feat/kafka-sharding-config

jjbayer commented on relay-server/src/actors/store.rs:
```diff
@@ -58,33 +58,134 @@ fn make_distinct_id(s: &str) -> Uuid {
+            Self::Single {
+                topic_name,
+                producer,
+            } => (topic_name.as_str(), Arc::clone(producer)),
```
nit: You can probably change the signature of `get_producer` to ` return (&str, &ThreadedProducer)` to save on `Arc::clone`s.

olksdr replied:
done in [1c3d5e0](https://github.com/getsentry/relay/pull/1454/commits/1c3d5e07e5a6d8352b6fadbed57a04d649374f0f)

untitaker commented on tests/integration/fixtures/relay.py:
```diff
@@ -133,7 +134,7 @@ def inner(
 
         dir = config_dir("relay")
-        dir.join("config.yml").write(json.dumps(default_opts))
+        dir.join("config.yml").write(yaml.dump(default_opts))
```
we don't have to change this now, but I believe that at some point the configuration will have to become json-serializable. then we will have to revisit the schema. but that's fine

olksdr committed:
ref: Address review comments
